lib/spack/spack/util/download.py

```diff
@@ -0,0 +1,381 @@
+            effective_url = self._response.geturl()
+        return RequestInfo(
+            url=self.url, effective_url=effective_url, headers=self._response.headers
+        )
+
+
+class CurlStreamReader(UrlStreamReader):
+    """Streams a file from a URL using curl."""
+
+    _curl_exe: Optional[str] = None
+
+    def __init__(
+        self,
+        *,
+        url: str,
+        timeout: int,
+        config_args: Optional[List[str]] = None,
+        cookie: Optional[str] = None,
+    ):
+        super().__init__(url=url, timeout=timeout)
+        self._cookie = cookie
+        self._config_args: List[str] = config_args or []
+        if CurlStreamReader._curl_exe is None:
+            CurlStreamReader._curl_exe = which_string("curl", required=True)
+        self._curl = CurlStreamReader._curl_exe
+
+    def __enter__(self) -> UrlStreamReader:
+        curl_args = (
+            base_curl_fetch_args(
+                self.url,
+                self.timeout,
+                headers=False,
+                status_bar=False,
+                user_agent=SPACK_USER_AGENT,
+            )
+            + self._redirect_header_args
+            + self._cookie_args
+            + self._config_args
+        )
+        curl_cmd = [self._curl] + curl_args
+
+        self._curl_process = subprocess.Popen(
+            curl_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        headers_parts = []
+
+        while True:
+            assert self._curl_process.stderr is not None, "curl process stderr is None"
+            line = self._curl_process.stderr.readline().decode("utf-8")
+            if line.strip() == "":
+                break
+            headers_parts.append(line)
+
+        headers = http.client.HTTPMessage()
+        scheme = urllib.parse.urlparse(self.url).scheme
+        if scheme in ("https", "http"):
+            try:
+                http_status = headers_parts[0].split()[1]
+                int(http_status)
+            except (IndexError, ValueError):
+                raise FetchError(f"Failed to fetch {self.url}: cannot parse HTTP status code")
+
+            if http_status.startswith("4") or http_status.startswith("5"):
+                status = HTTPStatus(int(http_status))
+                raise FetchError(f"Failed to fetch {self.url}: {status.value} {status.phrase}")
+
+        for line in headers_parts:
+            if ": " in line:
+                key, value = line.split(": ", 1)
+                headers[key.strip()] = value.strip()
+
+        self._headers = headers
+        self._progress = FetchProgress.from_headers(self._headers, enabled=sys.stdout.isatty())
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._progress.print(final=True)
+        self._curl_process.wait()
+        check_curl_code(self._curl_process.returncode)
+
+    def read(self, size: int) -> bytes:
+        assert self._curl_process.stdout is not None, "curl process stdout is None"
```

Comment:
this assertion shouldn't be in the hot loop

Reply:
I guess in a simpler implementation the assert would be outside of the loop

Reply:
Done